### ISSUE TITLE
Aggregation Strategy for Database

### ISSUE THREAD

**Issue by master12** [ONRR/nrrd#1451]
Figure out if there is a better way to get dynamic aggregations from our database

**Comment by Maroyafaied** [comment]
@mojobnichols please comment on this issue your strategy.

**Comment by mojobnichols** [comment]
The primary strategy will be to create a generalized view for a subject area (ie Revenue) with additional grouping columns (g1, g2, ...).  When a particular group is selected the query will add g1 to constraints of query (example g1='revenue_type') and go to that part of the view.   
the view will be a union of base views with possible g1 and g2 values and the corresponding grouping column will not be part of the base view group by server side.  

This should work well for home page graphs where there are no other filters.  In the case of query tool just using grouping columns might be sufficient server side aggregation, but if not we could introduce a concept of boolean filter columns (f1, f2, f3...) that would be true if that filter was selected.   If filter is not selected the query would use a view that had that filter rolled into it.   So if no state was selected or grouped on we could use a base view that had location rolled up and greatly reduce the amount of data returned.

The simpler form may be sufficient but the more complex form would always be going to the best level of aggregation possible, and in the case of selecting all filters it would go to highest level of detail but presumably the filters would result in a much better data set being returned. 

In conclusion the groups and filters will provide a map into a generalized view to the right level of aggregation.   The generalized view will be a view of sub-views at the right level of aggregation and can also be pre-computed through materialized views.
